File: js/src/decoder2/metadata/stream/StreamMetadata.ts
import { DecodingUtils } from '../../decoder/DecodingUtils';
import { IntWrapper } from '../../decoder/IntWrapper';
import { PhysicalStreamType } from './PhysicalStreamType';
import { LogicalStreamType } from './LogicalStreamType';
import { LogicalLevelTechnique } from './LogicalLevelTechnique';
import { PhysicalLevelTechnique } from './PhysicalLevelTechnique';
import { DictionaryType } from './DictionaryType';
import { OffsetType } from './OffsetType';
import { LengthType } from './LengthType';

export class StreamMetadata {
    private physical_stream_type: PhysicalStreamType;
    private logical_stream_type: LogicalStreamType;
    private logical_level_technique1: LogicalLevelTechnique;
    private logical_level_technique2: LogicalLevelTechnique;
    private physical_level_technique: PhysicalLevelTechnique;
    private num_values: number;
    private byte_length: number;

    constructor(
        physical_stream_type: PhysicalStreamType,
        logical_stream_type: LogicalStreamType,
        logical_level_technique1: LogicalLevelTechnique,
        logical_level_technique2: LogicalLevelTechnique,
        physical_level_technique: PhysicalLevelTechnique,
        num_values: number,
        byte_length: number
    ) {
        this.physical_stream_type = physical_stream_type;
        this.logical_stream_type = logical_stream_type;
        this.logical_level_technique1 = logical_level_technique1;
        this.logical_level_technique2 = logical_level_technique2;
        this.physical_level_technique = physical_level_technique;
        this.num_values = num_values;
        this.byte_length = byte_length;
    }

    public static decode(tile: Uint8Array, offset: IntWrapper): StreamMetadata {
        const stream_type = tile[offset.get()];
        const physical_stream_type = Object.values(PhysicalStreamType)[stream_type >> 4] as PhysicalStreamType;
        let logical_stream_type: LogicalStreamType | null = null;

        switch (physical_stream_type) {
            case PhysicalStreamType.DATA:
                logical_stream_type = new LogicalStreamType(Object.values(DictionaryType)[stream_type & 0xf] as DictionaryType);
                break;
            case PhysicalStreamType.OFFSET:
                logical_stream_type = new LogicalStreamType(null, Object.values(OffsetType)[stream_type & 0xf] as OffsetType);
                break;
            case PhysicalStreamType.LENGTH:
                logical_stream_type = new LogicalStreamType(null, null, Object.values(LengthType)[stream_type & 0xf] as LengthType);
                break;
        }
        offset.increment();

        const encodings_header = tile[offset.get()] & 0xFF;
        const logical_level_technique1 = Object.values(LogicalLevelTechnique)[encodings_header >> 5] as LogicalLevelTechnique;
        const logical_level_technique2 = Object.values(LogicalLevelTechnique)[encodings_header >> 2 & 0x7] as LogicalLevelTechnique;
        const physical_level_technique = Object.values(PhysicalLevelTechnique)[encodings_header & 0x3] as PhysicalLevelTechnique;
        offset.increment();

        const size_info = DecodingUtils.decodeVarint(tile, offset, 2);
        const num_values = size_info[0];
        const byte_length = size_info[1];

        return new StreamMetadata(
            physical_stream_type,
            logical_stream_type,
            logical_level_technique1,
            logical_level_technique2,
            physical_level_technique,
            num_values,
            byte_length
        );
    }

    public physicalStreamType(): PhysicalStreamType {
        return this.physical_stream_type;
    }

    public logicalStreamType(): LogicalStreamType {
        return this.logical_stream_type;
    }

    public logicalLevelTechnique1(): LogicalLevelTechnique {
        return this.logical_level_technique1;
    }

    public logicalLevelTechnique2(): LogicalLevelTechnique {
        return this.logical_level_technique2;
    }

    public physicalLevelTechnique(): PhysicalLevelTechnique {
        return this.physical_level_technique;
    }

    public numValues(): number {
        return this.num_values;
    }

    public byteLength(): number {
        return this.byte_length;
    }
}

File: java/src/main/java/org/maplibre/mlt/metadata/stream/StreamMetadata.java
package org.maplibre.mlt.metadata.stream;

import com.google.common.primitives.Bytes;
import org.maplibre.mlt.converter.encodings.EncodingUtils;
import org.maplibre.mlt.decoder.DecodingUtils;
import me.lemire.integercompression.IntWrapper;

public class StreamMetadata {
  private PhysicalStreamType physicalStreamType;
  private LogicalStreamType logicalStreamType;
  private LogicalLevelTechnique logicalLevelTechnique1;
  private LogicalLevelTechnique logicalLevelTechnique2;
  private PhysicalLevelTechnique physicalLevelTechnique;
  /* After logical Level technique was applied -> when rle is used it is the length of the runs and values array */
  private int numValues;
  private int byteLength;

  // TODO: refactor -> use builder pattern
  public StreamMetadata(
      PhysicalStreamType physicalStreamType,
      LogicalStreamType logicalStreamType,
      LogicalLevelTechnique logicalLevelTechnique1,
      LogicalLevelTechnique logicalLevelTechnique2,
      PhysicalLevelTechnique physicalLevelTechnique,
      int numValues,
      int byteLength) {
    this.physicalStreamType = physicalStreamType;
    this.logicalStreamType = logicalStreamType;
    this.logicalLevelTechnique1 = logicalLevelTechnique1;
    this.logicalLevelTechnique2 = logicalLevelTechnique2;
    this.physicalLevelTechnique = physicalLevelTechnique;
    this.numValues = numValues;
    this.byteLength = byteLength;
  }

  private int getLogicalType() {
    if (logicalStreamType == null) {
      return 0;
    }

    if (logicalStreamType.dictionaryType() != null) {
      return logicalStreamType.dictionaryType().ordinal();
    }

    if (logicalStreamType.lengthType() != null) {
      return logicalStreamType.lengthType().ordinal();
    }

    return logicalStreamType.offsetType().ordinal();
  }

  public byte[] encode() {
    var encodedStreamType = (byte) ((physicalStreamType.ordinal()) << 4 | getLogicalType());
    var encodedEncodingScheme =
        (byte)
            (logicalLevelTechnique1.ordinal() << 5
                | logicalLevelTechnique2.ordinal() << 2
                | physicalLevelTechnique.ordinal());
    var encodedLengthInfo =
        EncodingUtils.encodeVarints(new long[] {numValues, byteLength}, false, false);
    return Bytes.concat(new byte[] {encodedStreamType, encodedEncodingScheme}, encodedLengthInfo);
  }

  public static StreamMetadata decode(byte[] tile, IntWrapper offset) {
    var streamType = tile[offset.get()];
    var physicalStreamType = PhysicalStreamType.values()[streamType >> 4];
    LogicalStreamType logicalStreamType = null;
    switch (physicalStreamType) {
      case DATA:
        logicalStreamType = new LogicalStreamType(DictionaryType.values()[streamType & 0xf]);
        break;
      case OFFSET:
        logicalStreamType = new LogicalStreamType(OffsetType.values()[streamType & 0xf]);
        break;
      case LENGTH:
        logicalStreamType = new LogicalStreamType(LengthType.values()[streamType & 0xf]);
        break;
    }
    offset.increment();

    var encodingsHeader = tile[offset.get()] & 0xFF;
    var logicalLevelTechnique1 = LogicalLevelTechnique.values()[encodingsHeader >> 5];
    var logicalLevelTechnique2 = LogicalLevelTechnique.values()[encodingsHeader >> 2 & 0x7];
    var physicalLevelTechnique = PhysicalLevelTechnique.values()[encodingsHeader & 0x3];
    offset.increment();
    var sizeInfo = DecodingUtils.decodeVarint(tile, offset, 2);
    var numValues = sizeInfo[0];
    var byteLength = sizeInfo[1];

    return new StreamMetadata(
        physicalStreamType,
        logicalStreamType,
        logicalLevelTechnique1,
        logicalLevelTechnique2,
        physicalLevelTechnique,
        numValues,
        byteLength);
  }

  public PhysicalStreamType physicalStreamType() {
    return this.physicalStreamType;
  }

  public LogicalStreamType logicalStreamType() {
    return this.logicalStreamType;
  }

  public LogicalLevelTechnique logicalLevelTechnique1() {
    return this.logicalLevelTechnique1;
  }

  public LogicalLevelTechnique logicalLevelTechnique2() {
    return this.logicalLevelTechnique2;
  }

  public PhysicalLevelTechnique physicalLevelTechnique() {
    return this.physicalLevelTechnique;
  }

  public int numValues() {
    return this.numValues;
  }

  public int byteLength() {
    return this.byteLength;
  }
}
